python/lsst/daf/butler/registries/sqlRegistry.py

```diff
@@ -192,62 +193,101 @@ def makeDatabaseDict(self, table, types, key, value):
         config["table"] = table
         return SqlRegistryDatabaseDict(config, types=types, key=key, value=value, registry=self)
 
-    def _findDatasetId(self, collection, datasetType, dataId, **kwds):
-        """Lookup a dataset ID.
-
-        This can be used to obtain a ``dataset_id`` that permits the dataset
-        to be read from a `Datastore`.
+    def _makeDatasetRefFromRow(self, row, datasetType=None):
+        """Construct a DatasetRef from the result of a query on the Dataset
+        table.
 
         Parameters
         ----------
-        collection : `str`
-            Identifies the collection to search.
-        datasetType : `DatasetType` or `str`
-            A `DatasetType` or the name of one.
-        dataId : `dict` or `DataId`
-            A `dict` of `Dimension` link fields that label a Dataset
-            within a Collection.
-        kwds
-            Additional keyword arguments passed to the `DataId` constructor
-            to convert ``dataId`` to a true `DataId` or augment an existing
-            one.
+        row : `sqlalchemy.engine.RowProxy`.
+            Row of a query that contains all columns from the `Dataset` table.
+            May include additional fields (which will be ignored).
+        datasetType : `DatasetType`, optional
+            `DatasetType` associated with this dataset.  Will be retrieved
+            if not provided.
 
         Returns
         -------
-        dataset_id : `int` or `None`
-            ``dataset_id`` value, or `None` if no matching Dataset was found.
-
-        Raises
-        ------
-        ValueError
-            If dataId is invalid.
+        ref : `DatasetRef`.
+            A new `DatasetRef` instance.
         """
+        if datasetType is None:
```

Comment:
Some callers have an already-validated `DataseType` instance, and in that case, we want to just use it to avoid a database query (every little bit helps).  Other callers don't, and then we have no choice but to run the query.  We are just assuming that they don't conflict (i.e. that's a precondition for this function), and that's one reason this ought to stay a private function.  I'll document that more clearly.